src/c/gr_context.cpp

```diff
@@ -74,110 +85,184 @@ void gr_context_reset_context(gr_context_t* context, uint32_t state) {
 }
 
 gr_backend_t gr_context_get_backend(gr_context_t* context) {
-    return SK_ONLY_GPU_RETURN((gr_backend_t)AsGrContext(context)->backend(), (gr_backend_t)0);
+    return SK_ONLY_GPU((gr_backend_t)AsGrContext(context)->backend(), (gr_backend_t)0);
 }
 
 
 // GrGLInterface
 
 const gr_glinterface_t* gr_glinterface_create_native_interface() {
-    return SK_ONLY_GPU_RETURN(ToGrGLInterface(GrGLMakeNativeInterface().release()), nullptr);
+    return SK_ONLY_GPU(ToGrGLInterface(GrGLMakeNativeInterface().release()), nullptr);
 }
 
 const gr_glinterface_t* gr_glinterface_assemble_interface(void* ctx, gr_gl_get_proc get) {
-    return SK_ONLY_GPU_RETURN(ToGrGLInterface(GrGLMakeAssembledInterface(ctx, get).release()), nullptr);
+    return SK_ONLY_GPU(ToGrGLInterface(GrGLMakeAssembledInterface(ctx, get).release()), nullptr);
 }
 
 const gr_glinterface_t* gr_glinterface_assemble_gl_interface(void* ctx, gr_gl_get_proc get) {
-    return SK_ONLY_GPU_RETURN(ToGrGLInterface(GrGLMakeAssembledGLInterface(ctx, get).release()), nullptr);
+    return SK_ONLY_GPU(ToGrGLInterface(GrGLMakeAssembledGLInterface(ctx, get).release()), nullptr);
 }
 
 const gr_glinterface_t* gr_glinterface_assemble_gles_interface(void* ctx, gr_gl_get_proc get) {
-    return SK_ONLY_GPU_RETURN(ToGrGLInterface(GrGLMakeAssembledGLESInterface(ctx, get).release()), nullptr);
+    return SK_ONLY_GPU(ToGrGLInterface(GrGLMakeAssembledGLESInterface(ctx, get).release()), nullptr);
 }
 
 void gr_glinterface_unref(const gr_glinterface_t* glInterface) {
     SK_ONLY_GPU(SkSafeUnref(AsGrGLInterface(glInterface)));
 }
 
 bool gr_glinterface_validate(const gr_glinterface_t* glInterface) {
-    return SK_ONLY_GPU_RETURN(AsGrGLInterface(glInterface)->validate(), false);
+    return SK_ONLY_GPU(AsGrGLInterface(glInterface)->validate(), false);
 }
 
 bool gr_glinterface_has_extension(const gr_glinterface_t* glInterface, const char* extension) {
-    return SK_ONLY_GPU_RETURN(AsGrGLInterface(glInterface)->hasExtension(extension), false);
+    return SK_ONLY_GPU(AsGrGLInterface(glInterface)->hasExtension(extension), false);
 }
 
 
+// GrVkInstance
+
+gr_vkinterface_t* gr_vkinterface_make(vk_getinstanceprocaddr_t* vkGetInstanceProcAddr,
+                                      vk_getdeviceprocaddr_t* vkGetDeviceProcAddr,
+                                      vk_instance_t* vkInstance,
+                                      vk_device_t* vkDevice,
+                                      uint32_t extensionFlags) {
+    SK_ONLY_VULKAN(
```

Comment:
I am not really happy with this way of creating the interface. I would like to use the same format as `GRGlInterface` with the `Create` or `Assemble` static methods. This will allow for named overloads - something that can't be done with constructors. Also, I want to include the overloads that take delegates so it can be assembled via callbacks as that will be important for some platforms that do weird things.

Reply:
This is just how Vulkan is used. If you look at the Skia source they just do the same. The delegate approach might be something that is by the OpenGL design. It is very much possible to adopt this but I don't see why.